File: services/frontend/src/utils/htmlSanitizer.ts
/**
 * Sanitizador HTML compartilhado pelo admin (antes de salvar) e pelo storefront
 * (antes de renderizar via dangerouslySetInnerHTML). Espelha a whitelist do
 * backend (`app/core/html_sanitizer.py`) — manter ambos sincronizados.
 *
 * Estratégia: tags básicas para conteúdo (parágrafos, títulos, listas, links,
 * tabelas, imagens, span/div). Atributo `style` permitido com whitelist de
 * propriedades CSS (cor, tamanho, alinhamento, espaçamento) — flexibilidade
 * de customização sem abrir brecha para XSS via expression()/url(javascript:).
 */
import DOMPurify from 'dompurify'

export const ALLOWED_TAGS = [
  'p', 'br', 'hr',
  'strong', 'b', 'em', 'i', 'u', 's', 'sub', 'sup', 'mark', 'small',
  'ul', 'ol', 'li',
  'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
  'blockquote', 'code', 'pre',
  'a', 'img',
  'span', 'div',
  'table', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td',
]

export const ALLOWED_ATTRS = [
  'href', 'target', 'rel', 'title',
  'src', 'alt', 'width', 'height',
  'class', 'style',
  'colspan', 'rowspan',
]

// Propriedades CSS permitidas em `style`. Tudo fora dessa lista é descartado.
const ALLOWED_CSS_PROPS = new Set([
  'color', 'background-color', 'background',
  'font-size', 'font-weight', 'font-style', 'font-family',
  'text-align', 'text-decoration', 'text-transform', 'line-height', 'letter-spacing',
  'padding', 'padding-top', 'padding-right', 'padding-bottom', 'padding-left',
  'margin', 'margin-top', 'margin-right', 'margin-bottom', 'margin-left',
  'border', 'border-top', 'border-right', 'border-bottom', 'border-left',
  'border-color', 'border-style', 'border-width', 'border-radius',
  'width', 'height', 'max-width', 'min-width', 'max-height', 'min-height',
  'display', 'vertical-align',
])

// Padrões proibidos no valor de qualquer propriedade CSS.
const CSS_BAD_VALUE = /(expression|javascript:|vbscript:|data:(?!image\/)|@import|behavior|url\s*\()/i

function sanitizeStyleAttr(raw: string): string {
  return raw
    .split(';')
    .map(decl => decl.trim())
    .filter(Boolean)
    .map(decl => {
      const idx = decl.indexOf(':')
      if (idx < 0) return null
      const prop = decl.slice(0, idx).trim().toLowerCase()
      const val  = decl.slice(idx + 1).trim()
      if (!ALLOWED_CSS_PROPS.has(prop)) return null
      if (CSS_BAD_VALUE.test(val)) return null
      return `${prop}: ${val}`
    })
    .filter((d): d is string => d !== null)
    .join('; ')
}

let hookInstalled = false
function ensureHook() {
  if (hookInstalled) return
  DOMPurify.addHook('uponSanitizeAttribute', (_node, data) => {
    if (data.attrName === 'style' && typeof data.attrValue === 'string') {
      const cleaned = sanitizeStyleAttr(data.attrValue)
      if (!cleaned) { data.keepAttr = false; return }
      data.attrValue = cleaned
    }
    // Bloqueia URIs perigosas em href/src.
    if ((data.attrName === 'href' || data.attrName === 'src') && typeof data.attrValue === 'string') {
      const v = data.attrValue.trim().toLowerCase()
      if (v.startsWith('javascript:') || v.startsWith('vbscript:') ||
          (v.startsWith('data:') && !v.startsWith('data:image/'))) {
        data.keepAttr = false
      }
    }
  })
  // Força rel=noopener em links com target=_blank (anti tabnabbing).
  DOMPurify.addHook('afterSanitizeAttributes', (node) => {
    if (node.tagName === 'A' && (node as Element).getAttribute('target') === '_blank') {
      (node as Element).setAttribute('rel', 'noopener noreferrer')
    }
  })
  hookInstalled = true
}

/** Sanitiza HTML produzido pelo editor antes de persistir/renderizar. */
export function sanitizeHtml(dirty: string | null | undefined): string {
  if (!dirty) return ''
  ensureHook()
  return DOMPurify.sanitize(dirty, {
    ALLOWED_TAGS,
    ALLOWED_ATTR: ALLOWED_ATTRS,
    FORBID_TAGS: ['script', 'style', 'iframe', 'object', 'embed', 'form',
                  'input', 'button', 'select', 'textarea', 'link', 'meta', 'base'],
    FORBID_ATTR: ['onerror', 'onload', 'onclick', 'onmouseover', 'onfocus',
                  'onblur', 'onchange', 'onsubmit', 'srcdoc', 'formaction'],
    ALLOW_DATA_ATTR: false,
    KEEP_CONTENT: true,
  })
}

File: services/backend/app/core/html_sanitizer.py
"""
core/html_sanitizer.py
======================
Sanitização defensiva de HTML enviado pelo cliente para campos de conteúdo
rico (descrições de produtos, páginas, etc.). Whitelist espelhada com o
sanitizer do frontend (`utils/htmlSanitizer.ts`) — manter ambos sincronizados.

Política
--------
- Tags: estrutura básica (parágrafos, títulos, listas, link, imagem, tabela,
  span/div). Bloqueia script, iframe, style, form, etc.
- Atributos: href/target/rel, src/alt, class, style, colspan/rowspan.
- Style: somente propriedades CSS de apresentação (cor, tamanho, espaçamento).
  Bloqueia expression(), javascript:, url() e data: não-imagem.
- Links com target=_blank recebem rel=noopener noreferrer automaticamente.
"""
from __future__ import annotations

import re
from typing import Optional

import bleach
from bleach.css_sanitizer import CSSSanitizer

ALLOWED_TAGS = [
    "p", "br", "hr",
    "strong", "b", "em", "i", "u", "s", "sub", "sup", "mark", "small",
    "ul", "ol", "li",
    "h1", "h2", "h3", "h4", "h5", "h6",
    "blockquote", "code", "pre",
    "a", "img",
    "span", "div",
    "table", "thead", "tbody", "tfoot", "tr", "th", "td",
]

ALLOWED_ATTRS = {
    "*":   ["class", "style", "title"],
    "a":   ["href", "target", "rel"],
    "img": ["src", "alt", "width", "height"],
    "td":  ["colspan", "rowspan"],
    "th":  ["colspan", "rowspan"],
}

ALLOWED_CSS_PROPS = [
    "color", "background-color", "background",
    "font-size", "font-weight", "font-style", "font-family",
    "text-align", "text-decoration", "text-transform", "line-height", "letter-spacing",
    "padding", "padding-top", "padding-right", "padding-bottom", "padding-left",
    "margin", "margin-top", "margin-right", "margin-bottom", "margin-left",
    "border", "border-top", "border-right", "border-bottom", "border-left",
    "border-color", "border-style", "border-width", "border-radius",
    "width", "height", "max-width", "min-width", "max-height", "min-height",
    "display", "vertical-align",
]

ALLOWED_PROTOCOLS = ["http", "https", "mailto", "tel"]

_CSS_BAD_VALUE = re.compile(
    r"(expression|javascript:|vbscript:|@import|behavior|url\s*\()", re.IGNORECASE
)

_css_sanitizer = CSSSanitizer(allowed_css_properties=ALLOWED_CSS_PROPS)


def _post_filter_styles(html: str) -> str:
    """Remove declarações cujo valor casa com padrão perigoso (defesa em
    profundidade — bleach já bloqueia, mas reforçamos)."""
    def _scrub_style(match: re.Match[str]) -> str:
        decls = match.group(1).split(";")
        kept: list[str] = []
        for d in decls:
            d = d.strip()
            if not d:
                continue
            if _CSS_BAD_VALUE.search(d):
                continue
            kept.append(d)
        if not kept:
            return ""
        return f'style="{"; ".join(kept)}"'

    return re.sub(r'style="([^"]*)"', _scrub_style, html)


def sanitize_html(value: Optional[str]) -> Optional[str]:
    """Sanitiza HTML; devolve None se entrada for None, '' se ficar vazio."""
    if value is None:
        return None
    if not value.strip():
        return ""
    cleaned = bleach.clean(
        value,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRS,
        protocols=ALLOWED_PROTOCOLS,
        css_sanitizer=_css_sanitizer,
        strip=True,
        strip_comments=True,
    )
    cleaned = _post_filter_styles(cleaned)
    # Anti tab-nabbing: força rel=noopener em links target=_blank.
    cleaned = re.sub(
        r'(<a\b[^>]*\btarget="_blank"[^>]*?)(?<!\brel=")(>)',
        r'\1 rel="noopener noreferrer"\2',
        cleaned,
    )
    return cleaned


def sanitize_plain(value: Optional[str], max_len: Optional[int] = None) -> Optional[str]:
    """Strip total de tags — para campos meta (title, description) que devem
    ser texto puro. Trunca opcionalmente."""
    if value is None:
        return None
    plain = bleach.clean(value, tags=[], strip=True).strip()
    if max_len is not None and len(plain) > max_len:
        plain = plain[:max_len]
    return plain
